DATA B Jorge Angon & Priyanshu Rathor


https://www.kaggle.com/datasets/taweilo/washington-dc-historical-weather-20158202407
---



In [ ]:
from google.colab import files

In [ ]:
#import libraries
import numpy as nm
import pandas as pd
import matplotlib.pyplot as plt


In [ ]:
#Upload file
uploaded = files.upload()

Saving dc_weather[1].csv to dc_weather[1].csv


In [ ]:
df=pd.read_csv("dc_weather[1].csv")
df.head()

,name,datetime,tempmax,tempmin,temp,feelslikemax,feelslikemin,feelslike,dew,humidity,...,solarenergy,uvindex,severerisk,sunrise,sunset,moonphase,conditions,description,icon,stations
0,"Washington,DC,USA",2015-08-01,33.1,22.8,28.2,34.0,22.8,28.6,17.4,53.3,...,22.9,8,NaN,2015-08-01T06:08:52,2015-08-01T20:19:30,0.54,Partially cloudy,Partly cloudy throughout the day.,partly-cloudy-day,"KIAD,KDCA,72405013743,72403793728,72403093738,..."
1,"Washington,DC,USA",2015-08-02,32.0,22.8,27.3,31.1,22.8,27.2,15.4,49.1,...,22.4,8,NaN,2015-08-02T06:09:44,2015-08-02T20:18:29,0.58,Partially cloudy,Partly cloudy throughout the day.,partly-cloudy-day,"KIAD,KDCA,72405013743,72403793728,72403093738,..."
2,"Washington,DC,USA",2015-08-03,33.2,21.8,27.9,34.9,21.8,28.7,18.6,58.7,...,24.1,9,NaN,2015-08-03T06:10:38,2015-08-03T20:17:26,0.61,Partially cloudy,Partly cloudy throughout the day.,partly-cloudy-day,"KDCA,72405013743,72403793728,KADW,KDAA,7459401..."
3,"Washington,DC,USA",2015-08-04,35.3,24.9,29.3,36.6,24.9,30.4,19.0,55.2,...,21.9,9,NaN,2015-08-04T06:11:31,2015-08-04T20:16:22,0.65,"Rain, Partially cloudy",Partly cloudy throughout the day with late aft...,rain,"KDCA,72405013743,72403793728,KADW,KDAA,7459401..."
4,"Washington,DC,USA",2015-08-05,33.6,24.0,28.6,33.0,24.0,28.6,17.6,55.3,...,23.7,9,NaN,2015-08-05T06:12:24,2015-08-05T20:15:16,0.68,Partially cloudy,Clearing in the afternoon.,partly-cloudy-day,"KDCA,72405013743,72403793728,KGAI,KADW,KDAA,72..."


In [ ]:
#eliminate unnamed feature
df = df.loc[:, ~df.columns.str.contains("^Unnamed")]
df

,datetime,tempmax,tempmin,temp,feelslikemax,feelslikemin,feelslike,dew,humidity,precip,...,windspeed,winddir,sealevelpressure,cloudcover,visibility,solarradiation,solarenergy,uvindex,severerisk,moonphase
0,2015-08-01,33.1,22.8,28.2,34.0,22.8,28.6,17.4,53.3,0.000,...,15.9,260.4,1011.7,48.5,16.0,265.1,22.9,8,NaN,0.54
1,2015-08-02,32.0,22.8,27.3,31.1,22.8,27.2,15.4,49.1,0.000,...,22.3,248.3,1013.6,61.2,16.0,259.1,22.4,8,NaN,0.58
2,2015-08-03,33.2,21.8,27.9,34.9,21.8,28.7,18.6,58.7,0.000,...,24.8,186.0,1011.1,45.8,16.0,277.0,24.1,9,NaN,0.61
3,2015-08-04,35.3,24.9,29.3,36.6,24.9,30.4,19.0,55.2,1.231,...,22.6,232.8,1012.3,65.7,14.7,254.7,21.9,9,NaN,0.65
4,2015-08-05,33.6,24.0,28.6,33.0,24.0,28.6,17.6,55.3,0.000,...,21.6,315.9,1014.1,59.6,16.0,273.5,23.7,9,NaN,0.68
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3314,2024-08-27,33.3,22.5,27.7,33.7,22.5,27.9,18.9,62.4,0.000,...,17.0,139.4,1018.8,40.4,15.1,207.7,18.1,8,10.0,0.78
3315,2024-08-28,37.8,24.2,30.2,39.7,24.2,31.9,20.4,58.7,0.000,...,23.1,219.7,1016.2,70.3,16.0,194.8,16.9,7,10.0,0.81
3316,2024-08-29,32.8,23.3,28.1,38.1,23.3,30.3,21.9,69.7,12.655,...,23.3,60.9,1019.4,73.8,14.9,133.6,11.5,8,60.0,0.85
3317,2024-08-30,23.8,22.2,22.9,23.8,22.2,22.9,20.8,87.6,6.238,...,19.8,60.9,1022.0,99.2,12.8,39.6,3.4,2,10.0,0.88


In [ ]:
#Removing columns that are not required
drop_cols = ["name", "description", "icon", "stations", "conditions", "preciptype", "sunrise", "sunset"]
for c in drop_cols:
    if c in df.columns:
        df = df.drop(columns=[c])

In [ ]:
df

,datetime,tempmax,tempmin,temp,feelslikemax,feelslikemin,feelslike,dew,humidity,precip,...,windspeed,winddir,sealevelpressure,cloudcover,visibility,solarradiation,solarenergy,uvindex,severerisk,moonphase
0,2015-08-01,33.1,22.8,28.2,34.0,22.8,28.6,17.4,53.3,0.000,...,15.9,260.4,1011.7,48.5,16.0,265.1,22.9,8,NaN,0.54
1,2015-08-02,32.0,22.8,27.3,31.1,22.8,27.2,15.4,49.1,0.000,...,22.3,248.3,1013.6,61.2,16.0,259.1,22.4,8,NaN,0.58
2,2015-08-03,33.2,21.8,27.9,34.9,21.8,28.7,18.6,58.7,0.000,...,24.8,186.0,1011.1,45.8,16.0,277.0,24.1,9,NaN,0.61
3,2015-08-04,35.3,24.9,29.3,36.6,24.9,30.4,19.0,55.2,1.231,...,22.6,232.8,1012.3,65.7,14.7,254.7,21.9,9,NaN,0.65
4,2015-08-05,33.6,24.0,28.6,33.0,24.0,28.6,17.6,55.3,0.000,...,21.6,315.9,1014.1,59.6,16.0,273.5,23.7,9,NaN,0.68
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3314,2024-08-27,33.3,22.5,27.7,33.7,22.5,27.9,18.9,62.4,0.000,...,17.0,139.4,1018.8,40.4,15.1,207.7,18.1,8,10.0,0.78
3315,2024-08-28,37.8,24.2,30.2,39.7,24.2,31.9,20.4,58.7,0.000,...,23.1,219.7,1016.2,70.3,16.0,194.8,16.9,7,10.0,0.81
3316,2024-08-29,32.8,23.3,28.1,38.1,23.3,30.3,21.9,69.7,12.655,...,23.3,60.9,1019.4,73.8,14.9,133.6,11.5,8,60.0,0.85
3317,2024-08-30,23.8,22.2,22.9,23.8,22.2,22.9,20.8,87.6,6.238,...,19.8,60.9,1022.0,99.2,12.8,39.6,3.4,2,10.0,0.88


In [ ]:
df["datetime"] = pd.to_datetime(df["datetime"], errors="coerce")
df = df.dropna(subset=["datetime"])

In [ ]:
df

,datetime,tempmax,tempmin,temp,feelslikemax,feelslikemin,feelslike,dew,humidity,precip,...,windspeed,winddir,sealevelpressure,cloudcover,visibility,solarradiation,solarenergy,uvindex,severerisk,moonphase
0,2015-08-01,33.1,22.8,28.2,34.0,22.8,28.6,17.4,53.3,0.000,...,15.9,260.4,1011.7,48.5,16.0,265.1,22.9,8,NaN,0.54
1,2015-08-02,32.0,22.8,27.3,31.1,22.8,27.2,15.4,49.1,0.000,...,22.3,248.3,1013.6,61.2,16.0,259.1,22.4,8,NaN,0.58
2,2015-08-03,33.2,21.8,27.9,34.9,21.8,28.7,18.6,58.7,0.000,...,24.8,186.0,1011.1,45.8,16.0,277.0,24.1,9,NaN,0.61
3,2015-08-04,35.3,24.9,29.3,36.6,24.9,30.4,19.0,55.2,1.231,...,22.6,232.8,1012.3,65.7,14.7,254.7,21.9,9,NaN,0.65
4,2015-08-05,33.6,24.0,28.6,33.0,24.0,28.6,17.6,55.3,0.000,...,21.6,315.9,1014.1,59.6,16.0,273.5,23.7,9,NaN,0.68
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3314,2024-08-27,33.3,22.5,27.7,33.7,22.5,27.9,18.9,62.4,0.000,...,17.0,139.4,1018.8,40.4,15.1,207.7,18.1,8,10.0,0.78
3315,2024-08-28,37.8,24.2,30.2,39.7,24.2,31.9,20.4,58.7,0.000,...,23.1,219.7,1016.2,70.3,16.0,194.8,16.9,7,10.0,0.81
3316,2024-08-29,32.8,23.3,28.1,38.1,23.3,30.3,21.9,69.7,12.655,...,23.3,60.9,1019.4,73.8,14.9,133.6,11.5,8,60.0,0.85
3317,2024-08-30,23.8,22.2,22.9,23.8,22.2,22.9,20.8,87.6,6.238,...,19.8,60.9,1022.0,99.2,12.8,39.6,3.4,2,10.0,0.88


In [ ]:
#These are the values we are using to predict max tempreature values, you can easily find these values on your weather reports in your devices
cols_needed = ["tempmax", "tempmin", "humidity", "windspeed", "sealevelpressure", "precip"]
df = df[cols_needed]

In [ ]:
df

,tempmax,tempmin,temp,feelslikemax,feelslikemin,feelslike,dew,humidity,precip,precipprob,...,sealevelpressure,cloudcover,visibility,solarradiation,solarenergy,uvindex,severerisk,moonphase,month,day
0,33.1,22.8,28.2,34.0,22.8,28.6,17.4,53.3,0.000,0,...,1011.7,48.5,16.0,265.1,22.9,8,NaN,0.54,8,1
1,32.0,22.8,27.3,31.1,22.8,27.2,15.4,49.1,0.000,0,...,1013.6,61.2,16.0,259.1,22.4,8,NaN,0.58,8,2
2,33.2,21.8,27.9,34.9,21.8,28.7,18.6,58.7,0.000,0,...,1011.1,45.8,16.0,277.0,24.1,9,NaN,0.61,8,3
3,35.3,24.9,29.3,36.6,24.9,30.4,19.0,55.2,1.231,100,...,1012.3,65.7,14.7,254.7,21.9,9,NaN,0.65,8,4
4,33.6,24.0,28.6,33.0,24.0,28.6,17.6,55.3,0.000,0,...,1014.1,59.6,16.0,273.5,23.7,9,NaN,0.68,8,5
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3314,33.3,22.5,27.7,33.7,22.5,27.9,18.9,62.4,0.000,0,...,1018.8,40.4,15.1,207.7,18.1,8,10.0,0.78,8,27
3315,37.8,24.2,30.2,39.7,24.2,31.9,20.4,58.7,0.000,0,...,1016.2,70.3,16.0,194.8,16.9,7,10.0,0.81,8,28
3316,32.8,23.3,28.1,38.1,23.3,30.3,21.9,69.7,12.655,100,...,1019.4,73.8,14.9,133.6,11.5,8,60.0,0.85,8,29
3317,23.8,22.2,22.9,23.8,22.2,22.9,20.8,87.6,6.238,100,...,1022.0,99.2,12.8,39.6,3.4,2,10.0,0.88,8,30


In [ ]:
features = ["tempmin", "humidity", "windspeed", "sealevelpressure", "precip", "month"]

In [ ]:
X = df.drop(columns=["tempmax"])
y = df["tempmax"]

print("X shape:", X.shape)
print("y shape:", y.shape)



X shape: (3319, 5)
y shape: (3319,)


In [ ]:
#we used train test split divides the dataset into training and testing sets to evaluate model performance on unseen data
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
# Initialize and train a linear regression model
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(x_train, y_train)

LinearRegression()

In [ ]:
# Make predictions on the test set
y_pred = model.predict(x_test)

In [ ]:
#We use Mean Squared Error and R squared to get regression model performance and MSE measures prediction error and R squared measures how much variance the model explains.”
from sklearn.metrics import mean_squared_error, r2_score
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print("\n===== MODEL RESULTS =====")
print("Mean Squared Error:", mse)
print("R-squared:", r2)



===== MODEL RESULTS =====
Mean Squared Error: 9.440391629283859
R-squared: 0.896264818394489


In [ ]:
#we use this to get the output based onn the data we enter for prediction
while True:
    try:
        tempmin = float(input("Enter minimum temperature (tempmin): "))
        humidity = float(input("Enter humidity (%): "))
        windspeed = float(input("Enter windspeed: "))
        sealevelpressure = float(input("Enter sealevelpressure: "))
        precip = float(input("Enter precipitation amount: "))

        input_data = pd.DataFrame([[tempmin, humidity, windspeed, sealevelpressure, precip]],
                                  columns=["tempmin", "humidity", "windspeed", "sealevelpressure", "precip"])

        predicted_tempmax = model.predict(input_data)[0]
        print(f"\nPredicted Maximum Temperature (tempmax): {predicted_tempmax:.2f}")

        again = input("Predict again? (yes/no): ")
        if again.lower() != "yes":
            break

    except:
        print("Invalid input. Please enter numeric values only.")


Predicted Maximum Temperature (tempmax): 35.24


We used DC weather data to make a model that predicts the maximum temperature (tempmax) and we used different weather clues like minimum temperature, humidity, windspeed, pressure, and rain we tried other options but for that we need more deep understanding of weather and need some calulations through this anyone can use it. Firstly, we cleaned the data and removed columns we did not need and then we split the data into two parts: One part to train the model, One part to test the model. After training, we checked how good the model is using MSE and RSquared. Our R squared is about 0.896.This means the model explains almost 90% of the temperature changes. So the model is good. Then we made it interactive.
Now a user can enter weather values and the program will predict the maximum temperature.

In real life, this can help: Basic weather prediction, Planning outdoor events, Farming decisions, Energy planning, Business planning. It shows how machine learning can use many clues to make smart predictions.